extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduReader.java

```diff
@@ -0,0 +1,126 @@
+import org.locationtech.geowave.core.store.operations.ReaderParams;
+import org.locationtech.geowave.core.store.operations.RowReader;
+import org.locationtech.geowave.core.store.query.filter.ClientVisibilityFilter;
+import org.locationtech.geowave.core.store.util.DataStoreUtils;
+import org.locationtech.geowave.datastore.kudu.KuduRow;
+import org.locationtech.geowave.datastore.kudu.KuduRow.KuduField;
+import org.locationtech.geowave.datastore.kudu.operations.KuduOperations;
+import org.locationtech.geowave.datastore.kudu.util.KuduUtils;
+import org.locationtech.geowave.mapreduce.splits.RecordReaderParams;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Sets;
+import com.google.common.collect.Streams;
+
+public class KuduReader<T> implements RowReader<T> {
+	private final ReaderParams<T> readerParams;
+	private final RecordReaderParams recordReaderParams;
+	private final KuduOperations operations;
+	private final GeoWaveRowIteratorTransformer<T> rowTransformer;
+	private CloseableIterator<T> iterator;
+	private final boolean visibilityEnabled;
+	private static final Logger LOGGER = LoggerFactory.getLogger(KuduReader.class);
+
+	public KuduReader(final ReaderParams<T> readerParams, final KuduOperations operations,
+			final boolean visibilityEnabled) {
+		this.readerParams = readerParams;
+		recordReaderParams = null;
+		this.operations = operations;
+		this.rowTransformer = readerParams.getRowTransformer();
+		this.visibilityEnabled = visibilityEnabled;
+
+		initScanner();
+	}
+
+	@SuppressWarnings("unchecked")
+	protected void initScanner() {
+		try {
+			final Collection<SinglePartitionQueryRanges> ranges = readerParams.getQueryRanges()
+					.getPartitionQueryRanges();
+			if ((ranges != null) && !ranges.isEmpty()) {
+				iterator = operations.getKuduRangeRead(readerParams.getIndex().getName(), readerParams.getAdapterIds(),
+						ranges, DataStoreUtils.isMergingIteratorRequired(readerParams, visibilityEnabled),
+						rowTransformer,
+						new ClientVisibilityFilter(Sets.newHashSet(readerParams.getAdditionalAuthorizations())),
+						visibilityEnabled).results();
+			} else {
+				KuduTable table = operations.getTable(readerParams.getIndex().getName());
+				Schema schema = table.getSchema();
+				final List<RowResultIterator> results = new ArrayList<>();
+				Iterator<KuduRow> tmpIterator;
+				for (final short adapterId : readerParams.getAdapterIds()) {
+					KuduPredicate adapterIdPred = KuduPredicate.newComparisonPredicate(
+							schema.getColumn(KuduField.GW_ADAPTER_ID_KEY.getFieldName()), ComparisonOp.EQUAL,
+							adapterId);
+					KuduScannerBuilder scannerBuilder = operations.getScannerBuilder(table);
+					KuduScanner scanner = scannerBuilder.addPredicate(adapterIdPred).build();
+					while (scanner.hasMoreRows()) {
+						results.add(scanner.nextRows());
+					}
+					scanner.close();
+				}
+				if (visibilityEnabled) {
```

Comment:
There's a lot of duplicate code shared between this and `KuduRangeRead`. Perhaps we should refactor this into a common method.